server/models/procedures.sql

```diff
@@ -18,6 +18,235 @@ BEGIN
 END
 $$
 
+CREATE PROCEDURE StageInvoice(
+  IN is_distributable TINYINT(1),
+  IN date DATETIME,
+  IN cost DECIMAL(19, 4) UNSIGNED,
+  IN description TEXT,
+  IN service_id SMALLINT(5) UNSIGNED,
+  IN debtor_uuid BINARY(16),
+  IN project_id SMALLINT(5),
+  IN user_id SMALLINT(5),
+  IN uuid BINARY(16)
+)
+BEGIN 
+  DECLARE `no_invoice_stage` TINYINT(1) DEFAULT 0;
+  DECLARE CONTINUE HANDLER FOR SQLSTATE '42S02' SET `no_invoice_stage` = 1;
+  SELECT NULL FROM `stage_invoice` LIMIT 0;
+
+  IF (`no_invoice_stage` = 1) THEN 
+    create temporary table stage_invoice 
+    (select project_id, uuid, cost, debtor_uuid, service_id, user_id, date, description, is_distributable);
+  ELSE 
+    insert into stage_invoice
+    (select project_id, uuid, cost, debtor_uuid, service_id, user_id, date, description, is_distributable);
+  END IF;
+END $$
+
+CREATE PROCEDURE StageInvoiceItem(
+  IN uuid BINARY(16), 
+  IN inventory_uuid BINARY(16),
+  IN quantity INT(10) UNSIGNED, 
+  IN transaction_price decimal(19, 4),
+  IN inventory_price decimal(19, 4), 
+  IN debit decimal(19, 4), 
+  IN credit decimal(19, 4),
+  IN invoice_uuid BINARY(16)
+)
+BEGIN 
+  DECLARE `no_invoice_item_stage` TINYINT(1) DEFAULT 0;
+  DECLARE CONTINUE HANDLER FOR SQLSTATE '42S02' SET `no_invoice_item_stage` = 1;
+  SELECT NULL FROM `stage_invoice_item` LIMIT 0;
+
+  IF (`no_invoice_item_stage` = 1) THEN
+    -- tables does not exist - create and enter data
+    create temporary table stage_invoice_item 
+    (select uuid, inventory_uuid, quantity, transaction_price, inventory_price, debit, credit, invoice_uuid);
+  ELSE
+    -- table exists - only enter data
+    insert into stage_invoice_item
+    (select uuid, inventory_uuid, quantity, transaction_price, inventory_price, debit, credit, invoice_uuid);
+  END IF;
+END $$
+
+CREATE PROCEDURE StageBillingService(
+  IN id SMALLINT UNSIGNED,
+  IN invoice_uuid BINARY(16)
+)
+BEGIN 
+  CALL VerifyBillingServiceStageTable();
+
+   insert into stage_billing_service
+  (select id, invoice_uuid);
+END $$
+
+CREATE PROCEDURE StageSubsidy(
+  IN id SMALLINT UNSIGNED,
+  IN invoice_uuid BINARY(16)
+)
+BEGIN 
+  CALL VerifySubsidyStageTable();
```

Comment:
This is done for all optional entities, for these the staging tables cannot be guaranteed to exist and additional helper methods are used to ensure that errors are not thrown if no data for the staging process has been provided. 

I will add comments to reflect this.